spring-kafka/src/main/java/org/springframework/kafka/retrytopic/EndpointCustomizerFactory.java

```diff
@@ -0,0 +1,107 @@
+ *
+ * @author Tomaz Fernandes
+ * @since 2.7.2
+ *
+ * @see RetryTopicConfigurer
+ * @see DestinationTopic.Properties
+ * @see RetryTopicNamesProviderFactory.RetryTopicNamesProvider
+ *
+ */
+public class EndpointCustomizerFactory {
+
+	private final DestinationTopic.Properties destinationProperties;
+
+	private final EndpointHandlerMethod beanMethod;
+
+	private final BeanFactory beanFactory;
+
+	private final RetryTopicNamesProviderFactory retryTopicNamesProviderFactory;
+
+	EndpointCustomizerFactory(DestinationTopic.Properties destinationProperties, EndpointHandlerMethod beanMethod,
+							BeanFactory beanFactory, RetryTopicNamesProviderFactory retryTopicNamesProviderFactory) {
+
+		this.destinationProperties = destinationProperties;
+		this.beanMethod = beanMethod;
+		this.beanFactory = beanFactory;
+		this.retryTopicNamesProviderFactory = retryTopicNamesProviderFactory;
+	}
+
+	public EndpointCustomizer createEndpointCustomizer() {
+		return addSuffixesAndMethod(this.destinationProperties, this.beanMethod.resolveBean(this.beanFactory),
+				this.beanMethod.getMethod());
+	}
+
+	protected EndpointCustomizer addSuffixesAndMethod(DestinationTopic.Properties properties, Object bean, Method method) {
+		RetryTopicNamesProviderFactory.RetryTopicNamesProvider namesProvider =
+				this.retryTopicNamesProviderFactory.createRetryTopicNamesProvider(properties);
+		return endpoint -> {
+			Collection<EndpointCustomizer.TopicNamesHolder> topics = customizeAndRegisterTopics(namesProvider, endpoint);
+			endpoint.setId(namesProvider.getEndpointId(endpoint));
+			endpoint.setGroupId(namesProvider.getGroupId(endpoint));
+			endpoint.setTopics(topics.stream().map(EndpointCustomizer.TopicNamesHolder::getCustomizedTopic).toArray(String[]::new));
+			endpoint.setClientIdPrefix(namesProvider.getClientIdPrefix(endpoint));
+			endpoint.setGroup(namesProvider.getGroup(endpoint));
+			endpoint.setBean(bean);
+			endpoint.setMethod(method);
+			return topics;
+		};
+	}
+
+	protected Collection<EndpointCustomizer.TopicNamesHolder> customizeAndRegisterTopics(
+			RetryTopicNamesProviderFactory.RetryTopicNamesProvider namesProvider,
+			MethodKafkaListenerEndpoint<?, ?> endpoint) {
+
+		return getTopics(endpoint)
+				.stream()
+				.map(topic -> new EndpointCustomizer.TopicNamesHolder(topic, namesProvider.getTopicName(topic)))
+				.collect(Collectors.toList());
+	}
+
+	private Collection<String> getTopics(MethodKafkaListenerEndpoint<?, ?> endpoint) {
+		Collection<String> topics = endpoint.getTopics();
+		if (topics.isEmpty()) {
+			TopicPartitionOffset[] topicPartitionsToAssign = endpoint.getTopicPartitionsToAssign();
+			if (topicPartitionsToAssign != null && topicPartitionsToAssign.length > 0) {
+				topics = Arrays.stream(topicPartitionsToAssign)
+						.map(TopicPartitionOffset::getTopic)
+						.collect(Collectors.toList());
+			}
+		}
+
+		if (topics.isEmpty()) {
+			throw new IllegalStateException("No topics where provided for RetryTopicConfiguration.");
```

Comment:
Maybe we can give end-user more clues what is going on here, e.g. print those `destinationProperties` for context. This way, with a comprehensive exception message in the logs, end-user would be able to identify what exactly configuration is wrong.

Reply:
Also `were` not `where`.

Reply:
Good call. I think for the end user the properties themselves won't help much, so I added the endpoint method and bean class for context, let me know if you prefer otherwise.